File: mitopulse_universal_v1/mobile/ios/Sources/Crypto.swift
import Foundation
import CryptoKit

enum Crypto {
    static func b64url(_ data: Data) -> String {
        let s = data.base64EncodedString()
        return s.replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func hmacSHA256(secret: Data, msg: Data) -> Data {
        let key = SymmetricKey(data: secret)
        let mac = HMAC<SHA256>.authenticationCode(for: msg, using: key)
        return Data(mac)
    }

    static func dynamicId(secret: Data, canonicalJson: String) -> String {
        let mac = hmacSHA256(secret: secret, msg: Data(canonicalJson.utf8))
        return b64url(mac)
    }
}

File: mitopulse_universal_v1/mobile/android/src_snippets/Crypto.kt
package com.mitopulse.core

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Crypto {
    fun hmacSha256(secret: ByteArray, msg: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        return mac.doFinal(msg)
    }

    fun b64url(bytes: ByteArray): String =
        Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    fun dynamicId(secret: ByteArray, canonicalJson: String): String =
        b64url(hmacSha256(secret, canonicalJson.toByteArray(Charsets.UTF_8)))

    fun randomSecret(n: Int = 32): ByteArray {
        val b = ByteArray(n)
        SecureRandom().nextBytes(b)
        return b
    }
}
